### PR TITLE
Support custom validation schema

### PR DESCRIPTION
**Pull request by bluebill1049**
It's working in progress, but keep you guys in the loop. Feel free to add your suggestion or feedback.

### CODE REVIEW

**Comment by bluebill1049** [comment]
```
import React from "react";
import ReactDOM from "react-dom";
import useForm from "./src";

const Joi = require("@hapi/joi");

const schemaWhat = Joi.object({
  username: Joi.string()
    .alphanum()
    .min(3)
    .max(30)
    .required()
});

const resolver = values => {
  const { error, values: outcome } = schemaWhat.validate(values);

  return {
    values: outcome,
    errors: {
      test: ""
    } // map the error object here
  };
};

function App() {
  const { register, handleSubmit, errors } = useForm({
    validationSchema: {
      resolver
    }
  });
  
  return (
    <form onSubmit={handleSubmit(d => console.log(d))}>
      <input type="text" name="username" ref={register} />
    </form>
  );
}
```
Example code above.

File: package.json [hunk 1]
```diff
@@ -90,5 +90,9 @@
       "npm run prettier",
       "git add"
     ]
+  },
+  "dependencies": {
+    "@hapi/joi": "^16.1.7",
```
Comment (vatosarmat) [inline package.json]: React Hook Form is a tiny library without any dependencies

**Comment by bluebill1049** [comment]
closing it for now :)

**Comment by deklanw** [comment]
Is this still on the map?

**Comment by bluebill1049** [comment]
@deklanw it's on the map, but no one seems interested so parked :)

**Comment by JeromeDeLeon** [comment]
@bluebill1049 I'm really having a second thought about this before. I thought it would be useful but then, since using Yup, I haven't encountered any problems that want me to switch to the new one. 

@deklanw Could you provide use case/s where it is beneficial?

**Comment by deklanw** [comment]
@bluebill1049 @JeromeDeLeon 

Perhaps there's a better alternative to what I had in mind, if so I'm super open to it. I just want a way to share validation logic between frontend and backend, without using Yup, which I find bloated. Yup bundle size is about the bundle size of Formik and react-hook-form put together

Compare
https://bundlephobia.com/result?p=yup@0.28.0

with
https://bundlephobia.com/result?p=superstruct@0.8.3
https://bundlephobia.com/result?p=nope-validator@0.9.0

Maybe I'm missing something important that Yup does, but I haven't figured it out yet. I saw you were working on nope-validator support but the dev isn't responding https://github.com/react-hook-form/react-hook-form/issues/141

Is there some other way to share validation logic?

**Comment by JeromeDeLeon** [comment]
I haven't come accross with `nope-validator`. I'm not warry of the size on my end tho but seeing the bundle size is kinda make want to look for alternative and that's to merge this to master but then again, we have seen quite a few people that want this kind of feature, and we want to lower the feature involve if not so many users requested it. Will you reconsider this? @bluebill1049

**Comment by bluebill1049** [comment]
Thanks, @JeromeDeLeon & @deklanw 's response. I am happy to re-open that PR and let it sit there for some more time, see if any more dev wants such feature and we can go from there. sounds ok? @deklanw

**Comment by iceniveth** [comment]
+1 for custom validation schema.

**Comment by hemedani** [comment]
please support superstruct it's a very light lib for validation

**Comment by hemedani** [comment]
how can I vote to this pr ?

**Comment by bluebill1049** [comment]
@iceniveth & @hemedani vote here: https://github.com/react-hook-form/react-hook-form/pull/817

**Pull request closed without merging**